File: packages/ios/Sources/Components/Badge/Badge.swift
import SwiftUI

public struct KozmosBadge: View {
    let text: String
    let variant: KozmosBadgeVariant
    let size: KozmosBadgeSize
    let counter: String?
    let showCounter: Bool
    
    public enum KozmosBadgeVariant {
        case `default`
        case secondary
        case destructive
        case outline
        case ghost
        case link
        case success
    }

    public enum KozmosBadgeSize {
        case `default`
        case sm
        case lg
        case icon
    }
    
    public init(
        _ text: String,
        variant: KozmosBadgeVariant = .default,
        size: KozmosBadgeSize = .default,
        counter: String? = nil,
        showCounter: Bool = false
    ) {
        self.text = text
        self.variant = variant
        self.size = size
        self.counter = counter
        self.showCounter = showCounter
    }
    
    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
            Text(text)
                .font(.subheadline)
                .fontWeight(.medium)
                .underline(variant == .link)

            if hasCounter, let counter {
                KozmosCounter(counter, tone: counterTone)
            }
        }
            .padding(.horizontal, horizontalPadding)
            .padding(.vertical, 0)
            .frame(minWidth: size == .icon ? 44 : nil, minHeight: 44)
            .background(backgroundColor)
            .foregroundColor(foregroundColor)
            .cornerRadius(KozmosDimensions.primitivesLayoutRadius50)
            .overlay(
                RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius50)
                    .stroke(borderColor, lineWidth: variant == .outline ? 1 : 0)
            )
    }

    private var horizontalPadding: CGFloat {
        switch size {
        case .sm: return KozmosDimensions.primitivesLayoutSpacing150
        case .default: return KozmosDimensions.primitivesLayoutSpacing200
        case .lg: return KozmosDimensions.primitivesLayoutSpacing400
        case .icon: return 0
        }
    }

    private var backgroundColor: Color {
        switch variant {
        case .default:
            return KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
        case .secondary:
            return KozmosColors.componentsPrimaryButtonsNeutralButtonBackgroundIdle
        case .destructive:
            return KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
        case .success:
            return KozmosColors.semanticsDataTeal
        case .outline, .ghost, .link:
            return Color.clear
        }
    }

    private var foregroundColor: Color {
        switch variant {
        case .default:
            return KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
        case .secondary:
            return KozmosColors.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
        case .destructive:
            return KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle
        case .success:
            return KozmosColors.primitivesColorsForeground1000
        case .outline, .ghost, .link:
            return KozmosColors.componentsSecondaryButtonsThemedButtonForegroundContentIdle
        }
    }

    private var borderColor: Color {
        variant == .outline ? KozmosColors.componentsSecondaryButtonsThemedButtonForegroundContentIdle : Color.clear
    }

    private var hasCounter: Bool {
        size != .icon && showCounter && counter != nil
    }

    private var counterTone: KozmosCounterTone {
        switch variant {
        case .default, .destructive, .success:
            return .inverse
        case .secondary, .outline, .ghost, .link:
            return .neutral
        }
    }
}

File: packages/android/src/main/java/com/kozmos/components/Badge/Badge.kt
package com.kozmos.components.badge

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kozmos.components.counter.CounterTone
import com.kozmos.components.counter.KozmosCounter
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

enum class BadgeVariant {
    Default, Secondary, Destructive, Outline, Ghost, Link
}

enum class BadgeSize {
    Default, Sm, Lg, Icon
}

@Composable
fun KozmosBadge(
    text: String,
    variant: BadgeVariant = BadgeVariant.Default,
    modifier: Modifier = Modifier,
    size: BadgeSize = BadgeSize.Default,
    counter: String? = null,
    showCounter: Boolean = false
) {
    val (containerColor, contentColor) = when (variant) {
        BadgeVariant.Default -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonBackgroundIdle to KozmosThemeTokens.componentsPrimaryButtonsThemedButtonForegroundContentIdle
        BadgeVariant.Secondary -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonBackgroundIdle to KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
        BadgeVariant.Destructive -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonBackgroundIdle to KozmosThemeTokens.componentsPrimaryButtonsDangerButtonForegroundContentIdle
        BadgeVariant.Outline, BadgeVariant.Ghost, BadgeVariant.Link -> Color.Transparent to KozmosThemeTokens.componentsSecondaryButtonsThemedButtonForegroundContentIdle
    }

    val horizontalPadding = when (size) {
        BadgeSize.Sm -> KozmosDimensions.primitivesLayoutSpacing150
        BadgeSize.Default -> KozmosDimensions.primitivesLayoutSpacing200
        BadgeSize.Lg -> KozmosDimensions.primitivesLayoutSpacing400
        BadgeSize.Icon -> 0.dp
    }

    val hasCounter = size != BadgeSize.Icon && showCounter && counter != null

    Box(
        modifier = modifier
            .height(44.dp)
            .then(if (size == BadgeSize.Icon) Modifier.width(44.dp) else Modifier)
            .background(containerColor, RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
            .border(
                width = if (variant == BadgeVariant.Outline) 1.dp else 0.dp,
                color = if (variant == BadgeVariant.Outline) contentColor else Color.Transparent,
                shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)
            )
            .padding(horizontal = horizontalPadding, vertical = 0.dp),
        contentAlignment = Alignment.Center
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing50),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = text,
                color = contentColor,
                fontSize = 14.sp,
                fontWeight = FontWeight.Medium,
                textDecoration = if (variant == BadgeVariant.Link) TextDecoration.Underline else TextDecoration.None
            )

            if (hasCounter) {
                KozmosCounter(
                    text = counter.orEmpty(),
                    tone = badgeCounterTone(variant)
                )
            }
        }
    }
}

private fun badgeCounterTone(variant: BadgeVariant): CounterTone = when (variant) {
    BadgeVariant.Default,
    BadgeVariant.Destructive -> CounterTone.Inverse
    BadgeVariant.Secondary,
    BadgeVariant.Outline,
    BadgeVariant.Ghost,
    BadgeVariant.Link -> CounterTone.Neutral
}
